cou/apps/ovn.py

```diff
@@ -0,0 +1,94 @@
+from cou.apps.auxiliary_subordinate import OpenStackAuxiliarySubordinateApplication
+from cou.exceptions import ApplicationError
+from cou.steps import UpgradeStep
+from cou.utils.openstack import OpenStackRelease
+
+logger = logging.getLogger(__name__)
+
+
+@AppFactory.register_application(["ovn-central", "ovn-dedicated-chassis"])
+class OvnPrincipalApplication(OpenStackAuxiliaryApplication):
+    """Ovn central application class."""
+
+    def pre_upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Pre Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add pre upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        self._check_ovn_workload_version()
+        return [
+            self._get_upgrade_current_release_packages_plan(),
+            self._get_refresh_charm_plan(target),
+        ]
+
+    def _check_ovn_workload_version(self) -> None:
+        """Check whether is necessary to manually upgrade OVN to 22.03.
+
+        :raises ApplicationError: When workload version is lesser than 22.03.0.
+        """
+        if any((Version(unit.workload_version) < Version("22.03.0") for unit in self.units)):
+            raise ApplicationError(
+                (
+                    "It's recommended to upgrade OVN to 22.03 before upgrading the cloud. "
+                    "Follow the instructions at: "
+                    "https://docs.openstack.org/charm-guide/latest/project/procedures/"
+                    "ovn-upgrade-2203.html"
+                )
+            )
+
+
+@AppFactory.register_application(["ovn-chassis"])
+class OvnSubordinateApplication(OpenStackAuxiliarySubordinateApplication):
+    """Ovn chassis application class."""
+
+    def pre_upgrade_plan(self, target: OpenStackRelease) -> list[Optional[UpgradeStep]]:
+        """Pre Upgrade planning.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :return: Plan that will add pre upgrade as sub steps.
+        :rtype: list[Optional[UpgradeStep]]
+        """
+        self._check_ovn_workload_version()
+        return [
+            self._get_refresh_charm_plan(target),
+        ]
+
+    def _check_ovn_workload_version(self) -> None:
+        """Check whether is necessary to manually upgrade OVN to 22.03.
+
+        :raises ApplicationError:  When workload version is lesser than 22.03.0.
+        """
+        if Version(self.status.workload_version) < Version("22.03.0"):
+            raise ApplicationError(
+                (
+                    "It's recommended to upgrade OVN to 22.03 before upgrading the cloud. "
+                    "Follow the instructions at: "
+                    "https://docs.openstack.org/charm-guide/latest/project/procedures/"
+                    "ovn-upgrade-2203.html"
```

Comment:
Can we also mentioned that `{self.status.workload_version} is not currently handled` in addition to the recommandation?

Reply:
I've changed the description a little bit, but I preferred to not include `{self.status.workload_version}` because that represents just for the subordinates. For principal charms we would need to identify the unit with the version lower than 22.03 and I think we can be a little bit more generic to fit both cases.